### ISSUE TITLE
Spotty/Spotify-Connect: Essential firewall rules for the plugin to work are not described in the README.md

### ISSUE THREAD

**Issue by noothangpersonel** [michaelherger/librespot#11]
When your spotty plugin isn't running on a dedicated sound server like a Raspberry Pi, but rather on a home server, it is not a good practice to disable the firewall. I think it is good to be clear about which firewall rules are required. I now see spotty accepting connections on different port every time I restart (this is output from 'sudo netstat -tulpn |grep spotty' on my Ubuntu system):

`tcp    0    0 0.0.0.0:33583     0.0.0.0:*    LISTEN    26893/spotty-x86_64`
`udp    0    0 0.0.0.0:5353      0.0.0.0:*    LISTEN    26893/spotty-x86_64`
`udp6   0    0 :::5353           ::: *        LISTEN    26893/spotty-x86_64`

So that port 33583 changes each time LMS (and spotty) restarts.
The 5353 is for the zeroconf/avahi/mdns/bonjour protocol that is being used I guess(?)

Could you describe how to configure a firewall in the README.md? I can not safely open it up for all incoming ports. That is against most good practices for a home server.
Personally I'm happy with a description about the necessary rules just for linux (ufw then preferrably), but a generic description of the necessary rules for all systems seems smart as well.

If the rules can be safely altered upon installation of the plugin itself that would be fine with me as well, but I can imagine that's against some good practices.

Maarten

**Comment by michaelherger** [comment]
I'm sorry to say, I can't tell you. This also depends on what firewall you're using. AFAIK the underlying librespot (and Spotify) use mDNS to announce itself on the network. Maybe you can find some information for mDNS and your firewall?

**Comment by noothangpersonel** [comment]
Yeah that will be my focus next. Also I can imagine considering the nature of the protocol that it might be necessary to dynamically alter the firewall rules. I now do it using a fancy script, but I need to test first if this works safely for all usages. And maybe I can use a standard like NAT or NAPT or something to make sure this all can run solidly off-the-shelf and as platform independant as possible. I will keep in touch on any progress. But I guess we beter move this discussion to the librespot main branch, because it probably applies to most usages of that library using Spotify Connect

**Comment by noothangpersonel** [comment]
Meanwhile: can you describe how the IPC (interprocess communication) between your spotty binaries and the logitechmediaserver is done? Are you using tcp-ip (if yes: which protocol?) or some other way? I'm quite sure that works correctly, but I want to have the most complete picture about spotty, so I am able to identify all related network traffic for setting up the firewall correctly.

**Comment by michaelherger** [comment]
You really want the gory details? 😂 There are several channels

* Spotty (uppercase - the plugin) talks to the [Spotify web API](https://developer.spotify.com/documentation/web-api/reference/get-information-about-the-users-current-playback) to get player status information. It tries (badly) to keep volume, play/pause/stop etc. in sync.
* spotty (lowercase - the helper) talks to Spotify whatever secret sauce they use
* spotty talks to Spotty over the LMS JSON/RPC interface when it receives instructions from the Spotify backend to control playback
* Spotty/LMS launches a spotty singleton instance for every track to be played. This instance pipes the audio data into LMS' transcoding pipeline

**Comment by noothangpersonel** [comment]
Interesting: the debian package of the java version of librespot has an option to specify a specifc port it listens on for firewall purposes: https://github.com/spocon/spocon. It would really help me if the LMS Spotty plugin also would have such an input field in the settings page of the plugin, one port for each instance, instead of a random port

**Comment by michaelherger** [comment]
Oh, that's also available in librespot/spotty:

```
    -z, --zeroconf-port PORT
                        The port the internal server advertises over zeroconf
                        1 - 65535. Ports <= 1024 may require root privileges.
```

You'd have to see whether you can patch this in to Spotty somehow.

**Comment by noothangpersonel** [comment]
Exactly! ... Wait! what? What do you mean 'you'? :-)

**Comment by michaelherger** [comment]
You might want to add another `@helperArgs` entry for this before this line:

https://github.com/michaelherger/Spotty-Plugin/blob/e1ffe917047fbd98675ed6916ce64ba78b5d7e0f/Connect/Daemon.pm#L105

**Comment by noothangpersonel** [comment]
Need to get the Rust stuff running then first. I'm not rusty in Rust: I've never tried it before. Installing the dev environment gives me all kinds of warnings. It's an interesting language though I have to admit

**Comment by michaelherger** [comment]
No rust required: the option already is there. Spotty just doesn't expose or use it. But the link I gave is for the Spotty plugin side of things where you could potentially define the port.

**Comment by noothangpersonel** [comment]
Ah right, LMS plugins are in perl code , no change in the spotty helper then, just the Plugin

**Comment by noothangpersonel** [comment]
Not quite operational yet, but this is what I had in mind:
![image](https://github.com/michaelherger/librespot/assets/14052441/e73c65d0-bcc0-403d-a0e1-b036949336a4)

I'm sure you can decode it despite it being in Dutch

**Comment by michaelherger** [comment]
"Port" is quite international 😀. I personally don't see adding this to the distribution. That screen already is cluttered. And people tend to tweak whatever option they see, without understanding what they're doing.

Maybe I could build in the backend code to handle the values, but not provide the UI. That way people like you could edit the config file manually - or use their own little plugin to provide a UI for this setting?

**Comment by noothangpersonel** [comment]
Well yeah obviously I can use the internationalisation with your strings.txt, but ...

I do agree the UI should have as little as possible technical choices. Too many already indeed.

Maybe you can just reserve a much more limited range of ports. So if you just start counting from a fixed higher number (e.g.: 43290) and then if you have multiple players it will increase by 1 for each. Then an ordinary user can easily determine which ports to open on the firewall. There is no need for it to be random, is there?

**Comment by noothangpersonel** [comment]
If you're fine with the solution above I can try to implement it in a separate branch. You need to assign me this issue I guess? Can't create a branch at this moment

**Comment by michaelherger** [comment]
The ports would need to be dynamic to make sure they don't cause conflicts with other software using the same ports. That's why I don't want to have such a setting in the plugin. It's working fine for tens of thousands of users the way it is today.

**Comment by noothangpersonel** [comment]
Yeah you can be very proud of this product. It's really great. But it could be a bit of survivorship-bias there: you might have millions of users when this will be fixed :-)
But agreed: ofcourse when ports are not available you should pick the next one. That's how it works already right? But why does it need to be random? You could just add one until you got a free one. Then everybody is happy right?

**Comment by michaelherger** [comment]
That's questions you'd have to ask the librespot guys... I don't have enough insight into the innards to answer them.

**Comment by noothangpersonel** [comment]
The librespot guys will probably tell me to use the --zeroconf-port option which I should then change in my forked version of your LMS Spotty Plugin

Ok, I will stop bothering you. Thanks a lot for sharing your thoughts, and it was educational and enjoyable.

**Comment by noothangpersonel** [comment]
I think this issue is strongly related to this one: https://github.com/michaelherger/Spotty-Plugin/issues/26. Glad to see I haven't been the only one suffering from these issues. That vlan solution is something I will venture into as well, but I think still it can be done easier. I'm thinking of making a fork which does not have the ability to control it through the Squeezebox menu's, but just offers Spotify Connect functionality with discovery and without an account of it's own. Less configuration, less interfaces to implement, less potential for bugs

**Comment by michaelherger** [comment]
Right! There's two of you 😁. It's a pity the guy never followed up on the topic. Maybe you can ping him?

**Comment by noothangpersonel** [comment]
Well we're power users with a family subscription. The most enthusiastic spotify users ... but I think these are the kind of users that resort to solutions like LMS and spotty. But there are probably more of us, but some have probably decided to retire LMS entirely because of bugs like these. Which is a huge pitty, because I don't think there is anything around there with quite the same potential, ease-of-use and pricepoint for hardware 
Yeah I'm curious about his solutions, I might give him a ping

**Comment by michaelherger** [comment]
> Well we're power users with a family subscription. The most enthusiastic spotify users ...

same here

> But there are probably more of us, but some have probably decided to retire LMS entirely because of bugs like these. 

This is not a bug, is it? It's a missing feature. And I think I've given you enough information to find a way to "implement" that feature. Isn't it working?

**Comment by noothangpersonel** [comment]
I have no problem configuring the zeroconf-port anymore, but that still doesn't make it work for other accounts unfortunately

I managed to make any client with any premium account work flawlessly with spotifyd (the daemon version of librespot) (so without username/passwd configured in the daemon)
Therefore I opened:
* mdns (5353)
* zerconf port (I took the default: 1234)
* 57621 (default spotify port that needs to be opened for control just found that in a google search)
Then it works fine!

I need all three of these ports to be opened! ... but that is with spotifyd ...

But with your Spotty/spotty combination opening that 57621 port (https://www.speedguide.net/port.php?port=57621) does not do the trick ...

**Comment by michaelherger** [comment]
Can you start over describing what does work and what does not? Eg. can you authenticate Spotty using a Spotify app? 

And make sure each instance running has its own port which must be distinct from the others.

**Comment by noothangpersonel** [comment]
Won't have time to work on it at the moment. I'll get back to you soon

**Comment by noothangpersonel** [comment]
After reverting to your original version and my script that automatically detects and closes/opens the right Spotty zeroconf port(s), it all seems to work now, I must have messed something up along the way. And my comment about the 57621 port doesn't seem valid ... for now. Only time and a bit of messing around will tell if it's stable this way. Zeroconf will never be 'zero configuration' on a (home) server with firewall. But if it keeps working this way with my hacks, the problem has been identified and I'll fork your plugin to make the zeroconf a well documented 'lowconf' at least for debian or linux in general

**Comment by darblarg** [comment]
Ran across this same issue today. I'm running containers and the guides I used to build the LMS image specify the individual ports for LMS. In my case the fix was to to switch to host networking as I'm not directly exposed to the internet. If you tweak the language on the authorization screen to tell people people that spotify connect uses a dynamic port it could help people figure out a solution faster.

**Comment by michaelherger** [comment]
> If you tweak the language on the authorization screen to tell people people that spotify connect uses a dynamic port it could help people figure out a solution faster.

Would be great if you could submit a PR with this clarification.

**Comment by darblarg** [comment]
Alright I think I figured out how to do that. Apologize in advance if the etiquette is wrong as this is the first time I've submitted a PR.

**Comment by michaelherger** [comment]
It's all good, thanks!

**Comment by michaelherger** [comment]
Unfortunately Connect mode had to be removed. If it returns, then it'll be an all new implementation.

Thanks for your understanding.
